### PR TITLE
FINERACT-1992: Multi Delinquency event for installment range change

### PR DESCRIPTION
**Pull request by ruchiD**
## Description
Change made to emit range change event for installment level range changes.

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java [hunk 1]
```diff
@@ -506,6 +526,7 @@ private void setDelinquencyDetailsForInstallment(final Loan loan, final LoanRepa
                             installmentDelinquencyData.getDelinquentAmount());
                     loanInstallmentDelinquencyTagRepository.delete(previousInstallmentDelinquencyTag);
                     // event installment delinquency range change
+                    isDelinquencyRangeChanged = true;
                 } else {
                     previousInstallmentDelinquencyTag.setOutstandingAmount(installmentDelinquencyData.getDelinquentAmount());
```
Comment (reluxa) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java:531]: Shall we raise an event also if the ranges are not changed just the amount?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java [hunk 2]
```diff
@@ -398,6 +398,11 @@ private Map<String, Object> setLoanDelinquencyTag(Loan loan, Long delinquencyRan
                 loanDelinquencyTagPrev.setLiftedOnDate(transactionDate);
                 loanDelinquencyTagHistory.add(loanDelinquencyTagPrev);
                 changes.put("previous", loanDelinquencyTagPrev.getDelinquencyRange());
+                // event when loan goes out of delinquency we do not calculate at
+                // installment level and remove all installment tags, so event needs to raised here.
+                if (loan.isEnableInstallmentLevelDelinquency()) {
+                    businessEventNotifierService.notifyPostBusinessEvent(new LoanDelinquencyRangeChangeBusinessEvent(loan));
```
Comment (reluxa) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java:404]: Do we have tests for this?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 1 above)
Comment (ruchiD) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java:531]: @adamsaghy Please comment.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 2 above)
Comment (ruchiD) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java:404]: No, will add one.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 1 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java:531]: No need.

Comment (ruchiD) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java:531]: okay

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 2 above)
Comment (ruchiD) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java:404]: done.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java [hunk 3]
```diff
@@ -441,20 +450,29 @@ public int compare(DelinquencyRange o1, DelinquencyRange o2) {
...
             if (installmentsCollectionData.containsKey(installment.getId())) {
-                setInstallmentDelinquencyDetails(loan, installment, delinquencyBucket, installmentsCollectionData.get(installment.getId()));
+                boolean isDelinquencySetForInstallment = setInstallmentDelinquencyDetails(loan, installment, delinquencyBucket,
+                        installmentsCollectionData.get(installment.getId()));
+                isDelinquencyRangeChangedForAnyOfInstallment = isDelinquencyRangeChangedForAnyOfInstallment
+                        || isDelinquencySetForInstallment;
             }
         }
         // remove tags for non existing installments that got deleted due to re-schedule
         removeDelinquencyTagsForNonExistingInstallments(loan.getId());
```
Comment (reluxa) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java:463]: If an installment is deleted and we delete the related delinquency installment, then we might also want to raise an event, don't we?

Comment (ruchiD) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java:463]: These installments are non existing for the loan. Reschedule/Disbursal flow creates new installments for loan and deletes previous ones for loan, but tags for old non existing installments exits and needs to be cleaned up. No need to raise event.

**Review by reluxa — APPROVED** [review]
LGTM

**Pull request merged**